src/Tools/dotnet-monitor/Egress/S3/S3StorageEgressProvider.cs

```diff
@@ -0,0 +1,258 @@
+            }
+        }
+
+        private async Task<string> UploadToStorageAsync(
+            S3StorageEgressProviderOptions options,
+            Stream stream,
+            EgressArtifactSettings artifactSettings,
+            CancellationToken token)
+        {
+            try
+            {
+                var client = await CreateClientAsync(options, token);
+                if (stream.CanSeek) // use the stream directly
+                {
+                    var request = new PutObjectRequest
+                    {
+                        BucketName = options.BucketName,
+                        ContentType = artifactSettings.ContentType,
+                        Key = artifactSettings.Name,
+                        InputStream = stream,
+                        AutoCloseStream = false,
+                    };
+                    await client.PutObjectAsync(request, token);
+                }
+                else // copy temporary to memory stream locally
+                {
+                    using TransferUtility transferUtility = new(client);
+                    await transferUtility.UploadAsync(stream, options.BucketName, artifactSettings.Name, token);
+                }
+
+                string resourceId = GetResourceId(client, options, artifactSettings);
+                return resourceId;
+            }
+            catch (Exception e)
+            {
+                throw CreateException(string.Format(CultureInfo.CurrentCulture, Strings.ErrorMessage_EgressS3FailedDetailed, e.Message));
+            }
+        }
+
+        private string GetResourceId(IAmazonS3 client, S3StorageEgressProviderOptions options, EgressArtifactSettings artifactSettings)
+        {
+            if (!options.GeneratePresSignedUrl)
+                return $"BucketName={options.BucketName}, Key={artifactSettings.Name}";
+
+            DateTime expires = DateTime.UtcNow.AddMinutes(options.PreSignedUrlExpiryInMinutes);
+            string resourceId = client.GetPreSignedURL(new GetPreSignedUrlRequest { BucketName = options.BucketName, Key = artifactSettings.Name, Expires = expires });
+            Logger?.EgressProviderSavedStream(EgressProviderTypes.S3Storage, resourceId);
+            return resourceId;
+        }
+
+        private static async Task<IAmazonS3> CreateClientAsync(S3StorageEgressProviderOptions options, CancellationToken cancellationToken)
+        {
+            AWSCredentials awsCredentials = null;
+            AmazonS3Config configuration = new();
+            // use the specified access key and the secrets taken from configuration or a local file
+            if (!string.IsNullOrEmpty(options.AccessKeyId) && (!string.IsNullOrEmpty(options.SecretAccessKey) || !string.IsNullOrEmpty(options.SecretsAccessKeyFile)))
+            {
+                string secretAccessKeyId;
+                if (!string.IsNullOrEmpty(options.SecretsAccessKeyFile) && File.Exists(options.SecretsAccessKeyFile))
```

Comment:
Can you explain the scenario for this? This looks like a user a specify a file path to a file that contains the access key. The dotnet-monitor tool already has a way of doing this without specializing for this particular property; for example, if the user configured an S3Storage egress provider named `MyProvider`, then the user can create a file at `/etc/dotnet-monitor/Egress__MyProvider__SecretAccessKey` that contains the access key. See the [configuration sources](https://github.com/dotnet/dotnet-monitor/blob/main/documentation/configuration.md#configuration-sources) docs.

Reply:
That's true, but in case the file has to be located at some different place, there is no way to do that. So, I consider it useful to have this possibility of an optional setting.

Reply:
What's the case that the file _has_ to be written somewhere else? In Kubernetes, you can use config maps and/or secrets + projected volume or sub paths to mount combinations of configuration under `/etc/dotnet-monitor` or `/home/<user>/.config/dotnet-monitor` as key-per-file settings.

Reply:
This was not addressed but resolved without explanation. Either provide some documentation by AWS that this file _must_ be provided in a specific spot (such that a file path is required to be configured) OR remove this property entirely and rely on key-per-file configuration.